### PR TITLE
route classloads to classloader by package

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java [hunk 1]
```diff
@@ -77,6 +67,24 @@ public boolean hasLoadedClass(final String className) {
...
+    }
+    return super.loadClass(name);
+  }
+
+  String getPackageName(final String className) {
+    // intentionally not thread-safe: the worst case scenario is excess allocation/lookups
+    String packageName = this.cache.get();
+    if (null == packageName || !className.startsWith(packageName)) {
+      packageName = className.substring(0, className.lastIndexOf('.'));
+      this.cache = new WeakReference<>(packageName);
```
Comment (jbachorik) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java:83]: So, `cache` is caching the last requested `packageName`?

Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java:83]: yes, it could probably do with a rename and probably doesn't need to to be a weak reference any more (this contained a `Pair<String, ClassLoader>` at some point today)

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java [hunk 2]
```diff
@@ -1,42 +1,61 @@
...
-
           if (!entry.isDirectory() && entry.getName().startsWith(filePrefix)) {
             String name = entry.getName();
             // remove data suffix
             int end = name.endsWith(".classdata") ? name.length() - 4 : name.length();
-            filenameToEntry.put(name.substring(internalJarFileName.length(), end), entry);
+            String fileName = name.substring(internalJarFileName.length(), end);
+            filenameToEntry.put(fileName, entry);
+            if (fileName.endsWith(".class") && !fileName.startsWith(currentDir)) {
+              currentDir = fileName.substring(1, fileName.lastIndexOf('/'));
```
Comment (jbachorik) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java:55]: This should probably account for 'no-package' classes as well - eg. `lastIndexOf('/') == -1`

Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java:55]: yes, in several places, I'll handle these cases in the next commit.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java [hunk 3]
```diff
@@ -59,12 +83,26 @@ protected URLConnection openConnection(final URL url) throws IOException {
...
+        this.cache = new WeakReference<>(pair);
+      } else {
+        log.debug("{} not found in {}", filename, name);
+        throw notFound;
+      }
     } else {
-      throw new NoSuchFileException(url.getFile(), null, url.getFile() + " not in internal jar");
+      // hack: just happen to know this only ever happens twice,
+      // so dismiss cache after a hit
+      this.cache = NULL;
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java:106]: Do you think that on some JVMs we have been looking up each class **more** than twice?

**Comment by richardstartin** [comment]
I'm all for adding more tests but these aren't new edge cases; this PR **prevents** looking up classes multiple times per load in the vast majority of cases.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java:106]: I have no idea.  It was just a question based on the code.

[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/InternalJarURLHandler.java:106]: Let's say that it's different on some JVM - so what? Let's assume that on the other JVM that we only try to load once, then we never reach this condition. If we try to load three or more times every odd load after the second will be a bit slower and end up doing a lookup, but we probably have bigger problems in that case.

File: dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/DatadogClassLoaderTest.groovy [hunk 4]
```diff
@@ -49,4 +53,50 @@ class DatadogClassLoaderTest extends Specification {
...
+      new DatadogClassLoader.BootstrapClassLoaderProxy()
+    DatadogClassLoader shared = new DatadogClassLoader(testJarLocation, "parent", bootstrapProxy, null)
+    DatadogClassLoader.DelegateClassLoader child = new DatadogClassLoader.DelegateClassLoader(testJarLocation,
+      "child", bootstrapProxy, null, shared)
+
+    when:
+    Class<?> c = child.loadClass("a.b.c.C")
+
+    then:
+    c.getClassLoader() == shared
```
Comment (bantonsson) [inline dd-java-agent/agent-bootstrap/src/test/groovy/datadog/trace/bootstrap/DatadogClassLoaderTest.groovy:70]: A bit confusing that it's named `parent` in the jar and the test name, but `shared` in the code.
